README.md
# Kaltura Embed Code Generator

This Javascript library can be used to generate Kaltura Embed Codes.

It was designed to easiliy create embed codes on-the-fly for your projects.

## Quick Start

1. Include ```KalturaEmbedCodeGenerator.js``` to your page.

	```html
	<script src="/{path_to_script}/KalturaEmbedCodeGenerator.js"></script>
	```

2. Create a new instance of the generator, pass in configuration object and call ```getCode``` method.

	```html
	<textarea id="code"></textarea>
	<script>
	var gen = new kEmbedCodeGenerator({
		partnerId: 0000,
		uiConfId: 999,
		entryId: '0_xsd34g3'
	});
	document.getElementById('code').innerHTML = gen.getCode();
	</script>
	```

## Configuration

| Option     	| Description			| Default	|
| -------------	| ------------------------ 	| ------------- |
| embedType ```string``` | Embed code type to generate. Can we one of: 'auto', 'dynamic', 'thumb', 'iframe', 'legacy' 	| "auto" |
| playerId ```string``` |   The player element Id to generate | "kaltura_player" |
| protocol ```string``` | Embed HTTP protocol to use | "http" | 
| host ```string``` | Host for loading html5 library & kdp swf | "www.kaltura.com" | 
| securedHost ```string``` | Secured host for loading html5 library & kdp swf | "www.kaltura.com" | 
| widgetId ```string``` | Kaltura Widget Id | _{partnerId} | 
| partnerId ```int``` | Kaltura Partner Id | null | 
| cacheSt ```int``` | Add cacheSt parameter to bust cache. Should be future timestamp | null | 
| uiConfId ```int``` | Kaltura UiConf Id | null | 
| entryId ```string``` | Kaltura Entry Id | null | 
| entryMeta ```object``` | Entry Metadata object. Example: {name: "Foo", description: "Bar", thumbnailUrl: "http://foo.bar.com/img.jpg"} | null | 
| width ```int``` | Set player width | 400 | 
| height ```int``` | Set player height | 330 |
| attributes ```object``` | Adds additonal attributes to embed code. Example: { "class": "player" } | {} (empty object) |
| flashVars ```object``` | Adds flashVars to player. Example: { "autoPlay": "true" } | {} (empty object) |
| includeKalturaLinks ```boolean``` | Include Kaltura SEO links | true | 
| includeSeoMetadata ```boolean``` | Include Entry SEO Metadata ( taken from {entryMeta} option ) | false | 
| includeHtml5Library ```boolean``` | Inclde HTML5 Library in legacy embed type | true | 

You can set the configuration options when creating a new instance to set the defaults 
and you can also set the configuration when calling ```getCode``` method.

Any option that sets on ```getCode``` will override options that were set when creating the instance.

```javascript
var gen = new kEmbedCodeGenerator({
	'host': 'cdnapi.kaltura.com',
	'partnerId': 0000,
	'width': 680,
	'height': 420
});
console.log(gen.getCode({
	'uiConfId': 0000000,
	'entryId': 'x_0000000',	
	'height': 300 // Will generate embed with "height" of 300px 
}));
```

If you like to play with all the different options you're welcome to try out the [demo](http://kaltura.github.com/EmbedCodeGenerator/demo/)

## Dependencies

This library uses [Handlebars](http://handlebarsjs.com/) to manage the embed code templates and includes ```handlebars.runtime.js``` library.
If your project already uses handlebars, you can remove that file from ```grunt.js``` file.

This library also uses ```JSON.strinify``` method which is supported nativaliy in modern browsers.
If you need to support older browsers, please add the following code.

```html
<!--[if lt IE 9]>
<script src="http://ajax.cdnjs.com/ajax/libs/json2/20110223/json2.js"></script>
<![endif]-->
```

## Build minified version
Just run:
```
grunt
```
	


CHANGELOG.md
== 1.0.7 ( Mar, 16 2013 ) 
 - Added Id attribute to Iframe embed

== 1.0.6 ( Feb, 28 2013 )
 - Updated Iframe embed url

== 1.0.5 ( Feb, 27 2013 )
 - Fixed undefined cacheSt parameter
 
== 1.0.4 ( Feb, 26 2013 )
 - Fixed Auto embed with on page target ( missing width/height for target )

== 1.0.3 ( Feb, 21 2013 )
 - Fixed error in modifing the original params object in getCode method

== 1.0.2 ( Feb, 11th 2013 )
 - Added configuration for including Html5 library with legacy embed

== 1.0.1 ( Feb, 3rd 2013 )
 - Added support for Iframe embed
 - SEO support to all embed codes
 - Added Kaltura links to all embed codes

== 1.0.0 ( Feb, 2nd 2013 )
 - Initial release.


grunt.js
/*global module:false*/
module.exports = function(grunt) {

  // Project configuration.
  grunt.initConfig({
    meta: {
      version: '1.0.6',
      banner: '/*! Kaltura Embed Code Generator - v<%= meta.version %> - ' +
        '<%= grunt.template.today("yyyy-mm-dd") %>\n' +
        '* https://github.com/kaltura/EmbedCodeGenerator\n' +
        '* Copyright (c) <%= grunt.template.today("yyyy") %> ' +
        'Ran Yefet; Licensed MIT */'
    },
    handlebars: {
      all: {
        src: 'templates',
        dest: 'lib/templates.js'
      }
    },
    concat: {
      dist: {
        src: [
          '<banner:meta.banner>', 
          'lib/handlebars.runtime.js', 
          'lib/handlebars.helpers.js',
          'lib/templates.js', 
          'lib/utils.js',
          'lib/EmbedCodeGenerator.js'
        ],
        dest: 'dist/KalturaEmbedCodeGenerator.js'
      }
    },
    min: {
      dist: {
        src: ['<banner:meta.banner>', '<config:concat.dist.dest>'],
        dest: 'dist/KalturaEmbedCodeGenerator.min.js'
      }
    },
    watch: {
      files: '<config:lint.files>',
      tasks: 'lint qunit'
    },
    jshint: {
      options: {
        curly: true,
        eqeqeq: true,
        immed: true,
        latedef: true,
        newcap: true,
        noarg: true,
        sub: true,
        undef: true,
        boss: true,
        eqnull: true,
        browser: true
      },
      globals: {}
    },
    uglify: {}
  });

  // Add handlebars plugin
  grunt.loadNpmTasks('grunt-handlebars');
  // Default task.
  grunt.registerTask('default', 'handlebars concat min');

};


Gruntfile.js
/*global module:false*/
module.exports = function(grunt) {

  // Project configuration.
  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),
    meta: {
      banner: '/*! Kaltura Embed Code Generator - v<%= pkg.version %> - ' +
        '<%= grunt.template.today("yyyy-mm-dd") %>\n' +
        '* https://github.com/kaltura/EmbedCodeGenerator\n' +
        '* Copyright (c) <%= grunt.template.today("yyyy") %> ' +
        'Ran Yefet; Licensed MIT */\n'
    },
    handlebars: {
      compile: {
        options: {
          namespace: "Handlebars.templates"
        },
        files: {
          'lib/templates.js': 'templates/*'
        }
      }
    },
    concat: {
      options: {
        banner: '<%= meta.banner %>'
      },
      dist: {
        src: [
          'lib/handlebars.runtime.js', 
          'lib/handlebars.helpers.js',
          'lib/templates.js', 
          'lib/utils.js',
          'lib/EmbedCodeGenerator.js'
        ],
        dest: 'dist/<%= pkg.name %>.js'
      }
    },
    uglify: {
      options: {
        banner: '<%= meta.banner %>'
      },
      dist: {
        files: {
          'dist/<%= pkg.name %>.min.js': ['<%= concat.dist.dest %>']
        }
      }
    }
  });

  // Add grunt plugins
  grunt.loadNpmTasks('grunt-contrib-handlebars');
  grunt.loadNpmTasks('grunt-contrib-concat');
  grunt.loadNpmTasks('grunt-contrib-uglify');

  // Default task.
  grunt.registerTask('default', ['handlebars', 'concat', 'uglify:dist']);

};


demo/demo.js
(function(window, undefined) {

var Demo = {
	storageName: 'kEmbedCodeGeneratorConfig',

	escapeHTML: function( text ) {
    	return text.replace(/&/g,'&amp;').replace(/</g,'&lt;').replace(/>/g,'&gt;');
	},

	generateIframe: function(embedCode, container) {
		var container = document.getElementById(container);		
		var iframe = container.appendChild(document.createElement('iframe'));
		var newDoc = iframe.contentDocument;

		newDoc.open();
		newDoc.write( '<!doctype html><html><body>' + embedCode + '</body></html>' );
		newDoc.close();
	},

	setValue: function( $el, value ) {
		// Populate form fields
		switch( $el.prop('tagName') ) {
			case 'INPUT':
				switch( $el.prop('type') ) {
					case 'checkbox':
						$el.prop('checked', value);
					default:
						$el.val(value);
						break;
				}
				break;
			case 'SELECT':
				$.each($el.children(), function() {
					if( this.value == value ) {
						this.selected = true;
					}
				});
				break;
		}
	},

	addField: function( fieldName, key, value ) {
		var targetDiv = fieldName + 'Fields';
		var $row = $('<div />').addClass('row-fluid');
		var $remove = $('<div />')
						.addClass('span1 close')
						.html('X')
						.click(function(){
							$(this).parent().remove();
						});
		var $keyDiv = $('<div />').addClass('span5');
		var $valDiv = $('<div />').addClass('span5');
		var $hiddenInput = $('<input />').prop('type', 'hidden');
		var $keyInput = $('<input />')
						.prop({
							'class': 'input-medium',
							'placeholder': 'key'
						})
						.change(function() {
							$hiddenInput.prop('name', fieldName + '[' + this.value + ']');
						});
		var $valInput = $('<input />')
						.prop({
							'class': 'input-medium',
							'placeholder': 'value'
						})
						.change(function() {
							$hiddenInput.prop('value', this.value);
						});

		if( key && value !== undefined ) {
			$hiddenInput.prop({
				name: fieldName + '[' + key + ']',
				value: value
			});
			$keyInput.prop('value', key);
			$valInput.prop('value', value);
		}

		$keyDiv.append($keyInput);
		$valDiv.append($valInput);

		$row.append($hiddenInput, $remove, $keyDiv, $valDiv);
		$('#' + targetDiv).append($row);
	},

	populateFormFields: function( $form ) {
		var _this = this;
		$.each(this.getConfig(), function( key, val ) {

			if( key == 'flashVars' || key == 'attributes' ) {
				if( val && typeof val == 'object' ) {
					$.each(val, function( k, v ) {
						_this.addField( key, k, v );
					});
				}
				return true;
			}

			if( val && typeof val == 'object' ) {
				$.each(val, function( k, v ) {
					var $field = $form.find('[name="'+key+'[' + k + ']"]');
					_this.setValue( $field, v );
				});
			}
			else 
			{
				var $field = $form.find('[name=' + key + ']');
				_this.setValue( $field, val );
			}
		});
	},

	addBindings: function( $form ) {
		var _this = this;
		// Adds new field
		$('.add-field').click(function(e) {
			e.preventDefault();
			var fieldName = $(this).data('field')
			_this.addField(fieldName);
		});

		$('.reset').click(function() {
			if( localStorage && typeof localStorage.removeItem === 'function' ) {
				localStorage.removeItem(_this.storageName);
				location.reload();
			}		
		});
		$form.submit(function(e) {
			var config = $(this).serializeObject();
			config = _this.fixCheckboxes(config);
			// Save config to local storage
			if( localStorage && typeof localStorage.setItem === 'function' ) {
				localStorage.setItem(_this.storageName, JSON.stringify(config));
			}
			config = _this.parseConfig(config);
			try {
				var embedCode = _this.generator.getCode(config);
				$('code').html(_this.escapeHTML(embedCode));
				$('code').each(function(i, e) {hljs.highlightBlock(e)});
				$('#playerContainer').empty();
				_this.generateIframe(embedCode, 'playerContainer');
			} catch(err){
				alert(err);
			}
			return false;
		});		
	},

	fixCheckboxes: function( config ) {
		config.includeKalturaLinks = $('input[name=includeKalturaLinks]')[0].checked;
		config.includeSeoMetadata = $('input[name=includeSeoMetadata]')[0].checked;
		config.includeHtml5Library = $('input[name=includeHtml5Library]')[0].checked;
		return config;
	},

	parseConfig: function( config ) {
		if( ! config.flashVars ) {
			return config;
		}
		var flashVars = {},
			val;
		for( var key in config.flashVars ) {
			val = config.flashVars[key];
			if( val.charAt(0) == '{' && val.charAt(val.length) == '}') {
				val = JSON.parse(val)
			}
			flashVars[ key ] = val;
		}
		config.flashVars = flashVars;
		return config;
	},

	setConfig: function( config ) {
		var defaultConfig = this.generator.config();
		this.config = this.generator.extend(config, defaultConfig);
	},

	getConfig: function() {
		// Get data from local storage
		if( localStorage && typeof localStorage.getItem === 'function' ) {
			var userConfig = localStorage.getItem( this.storageName );
			if( userConfig ) {
				this.config = JSON.parse(userConfig);
			}
		}
		return this.config;
	},

	setup: function( config ) {
		var $form = $('#config');
		var _this = this;
		this.generator = new kEmbedCodeGenerator();
		this.setConfig(config);
		this.populateFormFields($form)
		this.addBindings($form);
		// Run the demo
		$form.submit();
	}
};

window.Demo = Demo;

})(this);

// jQuery.serializeObject
(function(e){e.fn.serializeObject=function(){var t=this,n={},r={},i={validate:/^[a-zA-Z][a-zA-Z0-9_]*(?:\[(?:\d*|[a-zA-Z0-9_]+)\])*$/,key:/[a-zA-Z0-9_]+|(?=\[\])/g,push:/^$/,fixed:/^\d+$/,named:/^[a-zA-Z0-9_]+$/};this.build=function(e,t,n){e[t]=n;return e};this.push_counter=function(e){if(r[e]===undefined){r[e]=0}return r[e]++};e.each(e(this).serializeArray(),function(){if(!i.validate.test(this.name)){return}var r,s=this.name.match(i.key),o=this.value,u=this.name;while((r=s.pop())!==undefined){u=u.replace(new RegExp("\\["+r+"\\]$"),"");if(r.match(i.push)){o=t.build([],t.push_counter(u),o)}else if(r.match(i.fixed)){o=t.build([],r,o)}else if(r.match(i.named)){o=t.build({},r,o)}}n=e.extend(true,n,o)});return n}})(jQuery);

lib/utils.js
// Add indexOf to array object
if (!Array.prototype.indexOf) {
    Array.prototype.indexOf = function (searchElement /*, fromIndex */ ) {
        "use strict";
        if (this == null) {
            throw new TypeError();
        }
        var t = Object(this);
        var len = t.length >>> 0;
        if (len === 0) {
            return -1;
        }
        var n = 0;
        if (arguments.length > 1) {
            n = Number(arguments[1]);
            if (n !== n) { // shortcut for verifying if it's NaN
                n = 0;
            } else if (n !== 0 && n !== Infinity && n !== -Infinity) {
                n = (n > 0 || -1) * Math.floor(Math.abs(n));
            }
        }
        if (n >= len) {
            return -1;
        }
        var k = n >= 0 ? n : Math.max(len - Math.abs(n), 0);
        for (; k < len; k++) {
            if (k in t && t[k] === searchElement) {
                return k;
            }
        }
        return -1;
    };
}
// Add keys for Object
if (!Object.keys) {
  Object.keys = (function () {
    var hasOwnProperty = Object.prototype.hasOwnProperty,
        hasDontEnumBug = !({toString: null}).propertyIsEnumerable('toString'),
        dontEnums = [
          'toString',
          'toLocaleString',
          'valueOf',
          'hasOwnProperty',
          'isPrototypeOf',
          'propertyIsEnumerable',
          'constructor'
        ],
        dontEnumsLength = dontEnums.length;
 
    return function (obj) {
      if (typeof obj !== 'object' && typeof obj !== 'function' || obj === null) {
        throw new TypeError('Object.keys called on non-object');
      }
 
      var result = [];
 
      for (var prop in obj) {
        if (hasOwnProperty.call(obj, prop)) {
          result.push(prop);
        }
      }
 
      if (hasDontEnumBug) {
        for (var i=0; i < dontEnumsLength; i++) {
          if (hasOwnProperty.call(obj, dontEnums[i])) {
            result.push(dontEnums[i]);
          } 
        }
      }
      return result;
    };
  })();
}

lib/handlebars.helpers.js
(function (window, Handlebars, undefined ) {
/**
* Transforms flashVars object into a string for Url or Flashvars string.
*
* @method flashVarsToUrl
* @param {Object} flashVarsObject A flashvars object
* @param {String} paramName The name parameter to add to url
* @return {String} Returns flashVars string like: &foo=bar or &param[foo]=bar
*/
var flashVarsToUrl = function( flashVarsObject, paramName ) {
	 var params = '';

	 var paramPrefix = (paramName) ? paramName + '[' : '';
	 var paramSuffix = (paramName) ? ']' : '';

	 for( var i in flashVarsObject ){
		 // check for object representation of plugin config:
		 if( typeof flashVarsObject[i] == 'object' ){
			 for( var j in flashVarsObject[i] ){
				 params+= '&' + paramPrefix + encodeURIComponent( i ) +
				 	'.' + encodeURIComponent( j ) + paramSuffix + 
				 	'=' + encodeURIComponent( flashVarsObject[i][j] );
			 }
		 } else {
			 params+= '&' + paramPrefix + encodeURIComponent( i ) + paramSuffix + '=' + encodeURIComponent( flashVarsObject[i] );
		 }
	 }
	 return params;
};

// Setup handlebars helpers
Handlebars.registerHelper('flashVarsUrl', function(flashVars) {
	return flashVarsToUrl(flashVars, 'flashvars');
});
Handlebars.registerHelper('flashVarsString', function(flashVars) {
	return flashVarsToUrl(flashVars);
});
Handlebars.registerHelper('elAttributes', function( attributes ) {
	var str = '';
	for( var i in attributes ) {
		str += ' ' + i + '="' + attributes[i] + '"';
	}
	return str;
});

// Include kaltura links
Handlebars.registerHelper('kalturaLinks', function() {
	if( ! this.includeKalturaLinks ) {
		return '';
	}
	var template = Handlebars.templates['templates/kaltura_links.hbs'];
	return template();
});

Handlebars.registerHelper('seoMetadata', function() {
	var template = Handlebars.templates['templates/seo_metadata.hbs'];
	return template(this);
});

})(this, this.Handlebars);

lib/EmbedCodeGenerator.js
(function( window, undefined ) {
/**
* Kaltura Embed Code Generator
* Used to generate different type of embed codes
* Depended on Handlebars ( http://handlebarsjs.com/ )
* 
* @class EmbedCodeGenerator
* @constructor
*/
var EmbedCodeGenerator = function( options ) {
	this.init( options );
};

EmbedCodeGenerator.prototype = {

	types: ['auto', 'dynamic', 'thumb', 'iframe', 'legacy'],
	required: ['widgetId', 'partnerId', 'uiConfId'],

	defaults: {
		/**
		* Embed code type to generate
		* Can we one of: ['auto', 'dynamic', 'thumb', 'iframe', 'legacy']
		* 
		* @property embedType
		* @type {String}
		* @default "auto"
		*/		
		embedType: 'auto',
		/**
		* The Player element Id / Name that will be used for embed code
		* 
		* @property playerId
		* @type {String}
		* @default "kaltura_player"
		*/			
		playerId: 'kaltura_player',
		/**
		* Embed HTTP protocol to use
		* Can we one of: ['http', 'https']
		* 
		* @property protocol
		* @type {String}
		* @default "http"
		*/			
		protocol: 'http',
		/**
		* Host for loading html5 library & kdp swf
		* 
		* @property host
		* @type {String}
		* @default "www.kaltura.com"
		*/			
		host: 'www.kaltura.com',
		/**
		* Secured host for loading html5 library & kdp swf
		* Used if protocol is: 'https'
		* 
		* @property securedHost
		* @type {String}
		* @default "www.kaltura.com"
		*/		
		securedHost: 'www.kaltura.com',
		/**
		* Kaltura Widget Id
		* 
		* @property widgetId
		* @type {String}
		* @default "_{partnerId}"
		*/
		widgetId: null,
		/**
		* Kaltura Partner Id
		* 
		* @property partnerId
		* @type {Number}
		* @default null,
		*/
		partnerId: null,
		/**
		* Add cacheSt parameter to bust cache
		* Should be unix timestamp of future time
		* 
		* @property cacheSt
		* @type {Number}
		* @default null,
		*/		
		cacheSt: null,
		/**
		* Kaltura UiConf Id
		* 
		* @property uiConfId
		* @type {Number}
		* @default null,
		*/		
		uiConfId: null,
		/**
		* Kaltura Entry Id
		* 
		* @property entryId
		* @type {String}
		* @default null,
		*/		
		entryId: null,		
		/**
		* Entry Object similar to:
		* {
		*	name: 'Foo', 
		*	description: 'Bar', 
		*	thumbUrl: 'http://cdnbakmi.kaltura.com/thumbnail/...'
		* }
		* 
		* @property entryMeta
		* @type {Object}
		* @default {},
		*/		
		entryMeta: {},
		/**
		* Sets Player Width
		* 
		* @property width
		* @type {Number}
		* @default 400,
		*/		
		width: 400,
		/**
		* Sets Player Height
		* 
		* @property height
		* @type {Number}
		* @default 330,
		*/		
		height: 330,
		/**
		* Adds additonal attributes to embed code.
		* Example:
		* {
		*	"class": "player"
		* }
		* 
		* @property attributes
		* @type {Object}
		* @default {},
		*/	
		attributes: {},
		/**
		* Adds flashVars to player
		* Example:
		* {
		*	"autoPlay": "true"
		* }
		* 
		* @property flashVars
		* @type {Object}
		* @default {},
		*/			
		flashVars: {},
		/**
		* Include Kaltura SEO links to embed code
		* 
		* @property includeKalturaLinks
		* @type {Boolean}
		* @default true,
		*/
		includeKalturaLinks: true,
		/**
		* Include Entry Seo Metadata
		* Metadata is taken from {entryMeta} object
		* 
		* @property includeSeoMetadata
		* @type {Boolean}
		* @default false,
		*/
		includeSeoMetadata: false,
		/**
		* Include HTML5 library script
		* 
		* @property includeHtml5Library
		* @type {Boolean}
		* @default true,
		*/
		includeHtml5Library: true
	},
	/**
	* Merge two object together
	*
	* @method extend
	* @param {Object} destination object to merge into
	* @param {Object} sourece object to merge from
	* @return {Object} Merged object
	*/
	extend: function(destination, source) {
	    for (var property in source) {
	        if (source.hasOwnProperty(property) && !destination.hasOwnProperty(property)) {
	            destination[property] = source[property];
	        }
	    }
	    return destination;
	},
	/**
	* Check if property is null
	*
	* @method isNull
	* @param {Any} property some var
	* @return {Boolean}
	*/
	isNull: function( property ) {
		if (property.length && property.length > 0) {
			return false;
		}
	    if (property.length && property.length === 0) {
	    	return true;
	    }
		if( typeof property === 'object' ) {
			return (Object.keys(property).length > 0) ? false : true;
		}
		return !property;
	},
	/**
	* Set default options to EmbedCodeGenerator instance
	*
	* @method init
	* @param {Object} options Configuration object based on defaults object
	* @return {Object} Returns the current instance
	*/
	init: function( options ) {

		options = options || {}; 

		var defaults = this.defaults;

		// Make sure Handlebars is available
		if( typeof Handlebars === undefined ) {
			throw 'Handlebars is not defined, please include Handlebars.js before this script';
		}

		// Merge options with defaults
		if( typeof options === 'object' ) {
			this.options = this.extend(options, this.defaults);
		}
		// Set widgetId to partnerId if not defined
		if( ! this.config('widgetId') && this.config('partnerId') ) {
			this.config('widgetId', '_' + this.config('partnerId'));
		}

		return this;
	},
	/**
	* Get or Set default configuration
	*
	* @method config
	* @param {String} key configuration property name
	* @param {Any} value to set
	* @return {Mixed} Return the value for the key, configuration object or null
	*/
	config: function( key, val ) {
		// Used as getter
		if( val === undefined && typeof key === 'string' && this.options.hasOwnProperty(key) ) {
			return this.options[ key ];
		}
		// Get all options
		if( key === undefined && val === undefined ) {
			return this.options;
		}
		// Used as setter
		if( typeof key === 'string' && val !== undefined ) {
			this.options[ key ] = val;
		}
		return null;
	},
	/**
	* Check if required parameters are missing
	*
	* @method checkRequiredParams
	* @param {Object} Configuration object
	* @return throws exception if missing parameters
	*/
	checkRequiredParams: function( params ) {
		var requiredLength = this.required.length,
			i = 0;
		// Check for required configuration
		for(i; i<requiredLength; i++) {
			if( this.isNull(params[this.required[i]]) ) {
				throw 'Missing required parameter: ' + this.required[i];
			}
		}
	},	
	/**
	* Check if embed type is part of types array
	*
	* @method checkValidType
	* @param {String} type - One of config embed types
	* @return throws exception if not valid
	*/
	checkValidType: function( type ) {
		var valid = (this.types.indexOf(type) !== -1) ? true : false;;
		if( !valid ) {
			throw 'Embed type: ' + type + ' is not valid. Available types: ' + this.types.join(",");
		}
	},
	/**
	* Get Handlebars template based on embed type
	*
	* @method getTemplate
	* @param {String} type - One of config embed types
	* @return {Mixed} If found returns Handlebars template function, else null
	*/
	getTemplate: function( type ) {
		// Dynamic embed and Thumb embed has the same template
		type = (type == 'thumb') ? 'dynamic' : type;
		var templateName = 'templates/' + type + '.hbs';
		return ( type && Handlebars.templates && Handlebars.templates[ templateName ] ) ? Handlebars.templates[ templateName ] : null;
	},
	/**
	* Check if embed type is using kWidget embed
	*
	* @method isKWidgetEmbed
	* @param {String} type - One of config embed types
	* @return {Boolean} true / false
	*/
	isKWidgetEmbed: function( type ) {
		return ( type == 'dynamic' || type == 'thumb' ) ? true : false;
	},
	/**
	* Get embed host based on protocol
	*
	* @method getHost
	* @param {Object} params Configuration object
	* @return {String} Embed host
	*/
	getHost: function( params ) {
		return (params.protocol === 'http') ? params.host : params.securedHost;
	},
	/**
	* Generate HTML5 library script url
	*
	* @method getScriptUrl
	* @param {Object} params Configuration object
	* @return {String} HTML5 library script Url
	*/
	getScriptUrl: function( params ) {
		return params.protocol + '://' + this.getHost(params) + '/p/' + params.partnerId + '/sp/' + params.partnerId + '00/embedIframeJs/uiconf_id/' + params.uiConfId + '/partner_id/' + params.partnerId;
	},
	/**
	* Generate Flash SWF url
	*
	* @method getSwfUrl
	* @param {Object} params Configuration object
	* @return {String} Flash player SWF url
	*/
	getSwfUrl: function( params ) {
		var cacheSt = (params.cacheSt) ? '/cache_st/' + params.cacheSt : '';
		var entryId = (params.entryId) ? '/entry_id/' + params.entryId : '';
		return params.protocol + '://' + this.getHost(params) + '/index.php/kwidget' + cacheSt + 
				'/wid/' + params.widgetId + '/uiconf_id/' + params.uiConfId + entryId;
	},
	/**
	* Generate attributes object based on configuration
	*
	* @method getAttributes
	* @param {Object} params Configuration object
	* @return {Object} Attributes object
	*/
	getAttributes: function( params ) {
		var attrs = {};

		// Add style attribute for dynamic / thumb embeds
		// Or if includeSeoMetadata is true
		if( this.isKWidgetEmbed( params.embedType ) || params.includeSeoMetadata ) {
			attrs['style'] = 'width: ' + params.width + 'px; height: ' + params.height + 'px;';
		}

		// Add Seo attributes
		if( params.includeSeoMetadata ) {
			if( params.embedType == 'legacy' ) {
				attrs["xmlns:dc"] = "http://purl.org/dc/terms/";
				attrs["xmlns:media"] = "http://search.yahoo.com/searchmonkey/media/";
				attrs["rel"] = "media:video";
				attrs["resource"] = this.getSwfUrl( params );
			} else {
				attrs['itemprop'] = 'video'; 
				attrs['itemscope itemtype'] = 'http://schema.org/VideoObject';
			}
		}

		return attrs;
	},
	/**
	* Generate kWidget object for HTML5 library
	*
	* @method getEmbedObject
	* @param {Object} params Configuration object
	* @return {Object} kWidget object
	*/
	getEmbedObject: function( params ) {
		// Used by kWidget.embed
		var embedObject = {
			targetId: params.playerId,		
			wid: params.widgetId,
			uiconf_id: params.uiConfId,
			flashvars: params.flashVars
		};
		// Add cacheSt
		if( params.cacheSt ) {
			embedObject['cache_st'] = params.cacheSt;
		}
		// Add entryId
		if( params.entryId ) {
			embedObject['entry_id'] = params.entryId;
		}
		// Transform object into a string
		return JSON.stringify(embedObject, null, 2);
	},
	/**
	* Generate Final Embed Code
	*
	* @method getCode
	* @param {Object} params Configuration object
	* @return {String} HTML embed code
	*/
	getCode: function( localParams ) {
		// Set default for params
		var params = (localParams === undefined) ? {} : this.extend({}, localParams);
		// Merge with options
		params = this.extend( params, this.config() );
		// Set widgetId to partnerId if undefined
		if( ! params.widgetId && params.partnerId ) {
			params.widgetId = '_' + params.partnerId;
		}

		this.checkRequiredParams(params); // Check for missing params
		this.checkValidType(params.embedType); // Check if embed type is valid

		// Check if we have a template
		var template = this.getTemplate(params.embedType);
		if( ! template ) {
			throw 'Template: ' + params.embedType + ' is not defined as Handlebars template';
		}

		// Add basic attributes for all embed codes
		var data = {
			host: this.getHost( params ),
			scriptUrl: this.getScriptUrl( params ),
			attributes: this.getAttributes( params )
		};
		// Add SWF Url for flash embeds
		if( params.embedType === 'legacy' ) {
			data['swfUrl'] = this.getSwfUrl( params );
		}
		// Add embed method and embed object for dynamic embeds
		if( this.isKWidgetEmbed( params.embedType ) ) {
			data['embedMethod'] = (params.embedType == 'dynamic') ? 'embed' : 'thumbEmbed';
			data['kWidgetObject'] = this.getEmbedObject( params );
		}

		data = this.extend( data, params );
		return template( data );
	}
};

// Export module to window object
window.kEmbedCodeGenerator = EmbedCodeGenerator;

 })(this);